File: engine/tare/flags.py
"""Uniswap v4 hook permission flags — the low 14 bits of the hook address.

Ported from LPLens (phases/05-hooks/flags.ts) and kept byte-identical to the TypeScript
package @tare/hookflags. Verified against the whole official registry: 8,974 bit
comparisons over 613 hooks, zero deviation.
"""

HOOK_FLAGS = {
    "BEFORE_INITIALIZE": 1 << 13,
    "AFTER_INITIALIZE": 1 << 12,
    "BEFORE_ADD_LIQUIDITY": 1 << 11,
    "AFTER_ADD_LIQUIDITY": 1 << 10,
    "BEFORE_REMOVE_LIQUIDITY": 1 << 9,
    "AFTER_REMOVE_LIQUIDITY": 1 << 8,
    "BEFORE_SWAP": 1 << 7,
    "AFTER_SWAP": 1 << 6,
    "BEFORE_DONATE": 1 << 5,
    "AFTER_DONATE": 1 << 4,
    "BEFORE_SWAP_RETURNS_DELTA": 1 << 3,
    "AFTER_SWAP_RETURNS_DELTA": 1 << 2,
    "AFTER_ADD_LIQUIDITY_RETURNS_DELTA": 1 << 1,
    "AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA": 1 << 0,
}

REGISTRY_NAME = {
    "BEFORE_INITIALIZE": "beforeInitialize",
    "AFTER_INITIALIZE": "afterInitialize",
    "BEFORE_ADD_LIQUIDITY": "beforeAddLiquidity",
    "AFTER_ADD_LIQUIDITY": "afterAddLiquidity",
    "BEFORE_REMOVE_LIQUIDITY": "beforeRemoveLiquidity",
    "AFTER_REMOVE_LIQUIDITY": "afterRemoveLiquidity",
    "BEFORE_SWAP": "beforeSwap",
    "AFTER_SWAP": "afterSwap",
    "BEFORE_DONATE": "beforeDonate",
    "AFTER_DONATE": "afterDonate",
    "BEFORE_SWAP_RETURNS_DELTA": "beforeSwapReturnsDelta",
    "AFTER_SWAP_RETURNS_DELTA": "afterSwapReturnsDelta",
    "AFTER_ADD_LIQUIDITY_RETURNS_DELTA": "afterAddLiquidityReturnsDelta",
    "AFTER_REMOVE_LIQUIDITY_RETURNS_DELTA": "afterRemoveLiquidityReturnsDelta",
}

MASK = (1 << 14) - 1

def decode_flags(addr: str):
    lo = int(addr, 16) & MASK
    return lo, [k for k, v in HOOK_FLAGS.items() if lo & v]

def can_alter_swap_output(addr: str) -> bool:
    lo, _ = decode_flags(addr)
    return bool(lo & (HOOK_FLAGS["BEFORE_SWAP_RETURNS_DELTA"] | HOOK_FLAGS["AFTER_SWAP_RETURNS_DELTA"]))

File: apps/api/src/assistant/flags.ts
/**
 * Les 14 permissions d'un hook v4.
 *
 * Elles ne sont stockees nulle part : elles SONT les 14 bits bas de l'adresse du hook,
 * ce qui est la raison d'etre du minage CREATE2 (Uniswap/v4-core/src/libraries/Hooks.sol).
 *
 * Ce fichier reprend la table de packages/hookflags/src/index.ts. Il la recopie au lieu
 * de l'importer parce que ce paquet n'est pas construit et que la chaine de build de
 * apps/api n'est pas de ce lot. La parite avec le registre officiel n'est pas supposee :
 * __tests__/flags.test.ts la verifie sur les 613 fiches, bit a bit.
 */
import type { FlagName } from "./actions.js";

export const HOOK_FLAG_BITS: Record<FlagName, number> = {
  beforeInitialize: 1 << 13,
  afterInitialize: 1 << 12,
  beforeAddLiquidity: 1 << 11,
  afterAddLiquidity: 1 << 10,
  beforeRemoveLiquidity: 1 << 9,
  afterRemoveLiquidity: 1 << 8,
  beforeSwap: 1 << 7,
  afterSwap: 1 << 6,
  beforeDonate: 1 << 5,
  afterDonate: 1 << 4,
  beforeSwapReturnsDelta: 1 << 3,
  afterSwapReturnsDelta: 1 << 2,
  afterAddLiquidityReturnsDelta: 1 << 1,
  afterRemoveLiquidityReturnsDelta: 1 << 0,
};

export const FLAG_ORDER = Object.keys(HOOK_FLAG_BITS) as FlagName[];

const MASK = (1 << 14) - 1;

export interface DecodedFlags {
  bitmap: number;
  active: FlagName[];
}

export function decodeFlags(hookAddress: string): DecodedFlags {
  const lo = Number(BigInt(hookAddress) & BigInt(MASK));
  return { bitmap: lo, active: FLAG_ORDER.filter((k) => (lo & HOOK_FLAG_BITS[k]) !== 0) };
}

export function hasFlag(hookAddress: string, flag: FlagName): boolean {
  return (decodeFlags(hookAddress).bitmap & HOOK_FLAG_BITS[flag]) !== 0;
}

/**
 * Un hook qui peut deplacer la sortie du swap est un hook qui peut y prendre.
 * C'est une CAPACITE declaree, pas une mesure : elle ne remplace jamais un bps.
 */
export function canAlterSwapOutput(hookAddress: string): boolean {
  const { bitmap } = decodeFlags(hookAddress);
  return (
    (bitmap &
      (HOOK_FLAG_BITS.beforeSwapReturnsDelta | HOOK_FLAG_BITS.afterSwapReturnsDelta)) !==
    0
  );
}
